src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs

```diff
@@ -81,12 +110,28 @@ public override void Fail(string message)
         /// </devdoc>
         public override void Fail(string message, string detailMessage)
         {
-            // UIAssert is not enabled.
-            WriteAssert(string.Empty, message, detailMessage);
+            StackTrace stack = new StackTrace(true);
+            string stackTrace;
+            try
+            {
+                stackTrace = stack.ToString();
+            }
+            catch
+            {
+                stackTrace = "";
+            }
+            WriteAssert(stackTrace, message, detailMessage);
+            if (AssertUiEnabled)
+            {
+                s_provider.ShowDialog(stackTrace, message, detailMessage, "Assertion Failed");
+            }
+            else if (Debugger.IsAttached)
+                Debugger.Break();
```

Comment:
Please don't mix braces and no braces in if/else. It's harder to read and easier to make a mistake in maintenance.